### PR TITLE
Implements IAsyncParseNodeFactory interface

### PR DESCRIPTION
**Pull request by MihaMarkic**
Adds async support and marks sync method obsolete

### CODE REVIEW

File: src/JsonParseNodeFactory.cs [hunk 1]
```diff
@@ -56,5 +59,25 @@ public IParseNode GetRootParseNode(string contentType, Stream content)
...
+        /// <returns>An instance of <see cref="IParseNode"/> for json manipulation</returns>
+        public async Task<IParseNode> GetRootParseNodeAsync(string contentType, Stream content, 
+            CancellationToken cancellationToken = default) 
+        {
+            if(string.IsNullOrEmpty(contentType))
+                throw new ArgumentNullException(nameof(contentType));
+            else if(!ValidContentType.Equals(contentType, StringComparison.OrdinalIgnoreCase))
+                throw new ArgumentOutOfRangeException($"expected a {ValidContentType} content type");
+
+            _ = content ?? throw new ArgumentNullException(nameof(content));
```
Comment (baywet) [inline src/JsonParseNodeFactory.cs:77]: all the defensive programing should be moved to a private method, and reused across the public methods

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for putting this together! a minor recommendation

File: src/JsonParseNodeFactory.cs (hunk 1 above)
Comment (MihaMarkic) [reply src/JsonParseNodeFactory.cs:77]: Actually I was considering that - but it's just checking two arguments and a field - moving it to a separate method would make it less obvious and they sync method will eventually be removed anyway. You still want to move it to a separate method?

Comment (baywet) [reply src/JsonParseNodeFactory.cs:77]: I don't have a strong opinion on the matter. Let's proceed to a vote. @andrueastman ?

Comment (andrueastman) [reply src/JsonParseNodeFactory.cs:77]: I try to use [3 as a threshold](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) for these scenarios. So, we can just keep it as it is currently.

**Review by andrueastman — DISMISSED** [review]

**Review by baywet — DISMISSED** [review]

**Comment by baywet** [comment]
@MihaMarkic is this good to merge from your perspective?

File: src/JsonParseNodeFactory.cs [hunk 2]
```diff
@@ -56,5 +59,25 @@ public IParseNode GetRootParseNode(string contentType, Stream content)
...
+
+            using var jsonDocument = await JsonDocument.ParseAsync(content, cancellationToken: cancellationToken);
```
Comment (andrueastman) [inline src/JsonParseNodeFactory.cs:79]:
```suggestion
            using var jsonDocument = await JsonDocument.ParseAsync(content, cancellationToken: cancellationToken).ConfigureAwait(false);
```

**Review by baywet — APPROVED** [review]

**Comment by MihaMarkic** [comment]
@baywet Yep

**Review by andrueastman — APPROVED** [review]

**Pull request merged**